src/or/hs_service.c

```diff
@@ -1031,6 +1034,222 @@ load_service_keys(hs_service_t *service)
+  client_keys_dir_path = hs_path_from_filename(config->directory_path,
+                                               dir_client_seckeys);
+  if (BUG(hs_check_service_private_dir(get_options()->User,
+                                       client_keys_dir_path,
+                                       config->dir_group_readable, 1) < 0)) {
+    goto end;
+  }
+
+  /* If the client authorization is enabled, we must have a list of clients */
+  tor_assert(config->clients);
+
+  /* Load public keys from the file. If the file does not exist,
+   * parsed_clients will be empty. */
+  fname = hs_path_from_filename(config->directory_path,
+                                        fname_client_pubkeys);
+  client_keys_str = read_file_to_str(fname,
+                                     RFTS_IGNORE_MISSING,
+                                     &file_stat);
+  if (client_keys_str) {
+    if (parse_client_keys(parsed_clients,
+                          client_keys_str,
+                          client_keys_str + file_stat.st_size) < 0) {
+      log_warn(LD_REND, "Previously stored client key file could not "
+                        "be parsed.");
+      goto end;
+    } else {
+      log_info(LD_REND, "Parsed %d previously stored client entries.",
+               strmap_size(parsed_clients));
+    }
+  }
+
+  /* Either use loaded keys for configured clients or generate new
+   * ones if a client is new. */
+  SMARTLIST_FOREACH_BEGIN(config->clients,
+                          hs_service_authorized_client_t *, client) {
+    int pubkey_entry_len;
+    char *pubkey_entry = NULL;
+    char *fname_seckey = NULL;
+    hs_service_authorized_client_t *parsed =
+      strmap_get(parsed_clients, client->client_name);
+
+    if (parsed) {
+      log_info(LD_REND, "The public key for client %s already exists."
+                        "Loading the key.", client->client_name);
+
+      memcpy(&client->client_pk,
+             &parsed->client_pk,
+             sizeof(curve25519_secret_key_t));
+    } else {
+      curve25519_secret_key_t client_seckey;
+      log_info(LD_REND, "The public key for client %s does not exist."
+                        "Generating a new key.", client->client_name);
+      if (curve25519_secret_key_generate(&client_seckey, 0) < 0) {
+        log_warn(LD_REND, "Client secret key cannot be generated.");
+        goto end;
+      }
+
+      curve25519_public_key_generate(&client->client_pk, &client_seckey);
+
+      /* Write a newly generated secret key to a file so that the service
+       * can send this secret key to the client in a secure out-of-band
+       * way. */
+      fname_seckey = client_seckey_path_from_name(config->directory_path,
+                                                  client->client_name);
+      /* We do not need to check the return value because the secret key
+       * was appropriately generated */
+      curve25519_secret_to_base64(buf, &client_seckey);
```

Comment:
I also wonder if we should try to prefix secret keys with some strings so that we can recognize them as client auth private keys. And same for public keys.

I'm kind afraid of HS operators accidentally sending public keys to their clients instead of private keys, since they both have the same format, and the clients' authentication would fail without a reason...

Another way, to make this less likely, is to add some sort of header on top of the public/private key files to explain the functionality of those files.

What do you think?

Reply:
Hmm. I wonder if we should first figure out the whole UX of client authorization before figuring out the prefix here. That is, perhaps we should figure out how the control port interface UX should work etc. to see whether a prefix is useful or not... 

Also, I have mixed feelinsg about `curve25519-private-key.[base64]` mainly because it's big and also the dash and the dot makes it harder to copy-paste.

Need to think more. Perhaps a [tor-dev] post is in order.

Reply:
This `BIP` specifies how bitcoin is prefixing their keys: https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki

They are using base32 tho, instead of base64. I wonder if that's gonna make our keys grow too huge, like our addresses...